fees: make the class FeeFilterRounder thread-safe

vasild opened the pull request:
Make the class `FeeFilterRounder` thread-safe so that its methods can be called concurrently by different threads on the same object. Currently it has just one method (`round()`).

The second commit is optional, but it improves readability, showing that the `feeset` member will never be changed, thus does not need protection from concurrent access.

vasild committed:
fees: make the class FeeFilterRounder thread-safe

So that its methods can be called concurrently by different threads on
the same object. Currently it has just one method (`round()`).

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

vasild committed:
fees: make FeeFilterRounder::feeset const

It is only set in the constructor, thus improve readability by marking
it as `const` and setting it from the initializer list using a helper
function to derive its value.

The idea was suggested by Anthony Towns <aj@erisian.com.au> in
https://github.com/bitcoin/bitcoin/pull/19268#discussion_r439929792

MarcoFalke commented:
> This achieves the same purpose as https://github.com/bitcoin/bitcoin/pull/22053, but in a different way (cc @MarcoFalke).

I don't think the two pulls achieve the same goal. In fact I think they are completely orthogonal:

* This pull makes a utility class thread safe, so that it may be called in a concurrent context without running into UB
* The other pull aims to reduce the use of the `cs_main` "global" consistency lock. The consistency lock is needed even in the absence of UB. Also, `m_insecure_rand_mutex` added here can't possibly achieve what `cs_main` achieves right now in `MaybeSendFeefilter`.

So I think the two pull requests are best reviewed and evaluated independently. Also, there is no dependency between them. Either can be merged before the other, or not at all.

jonatack commented on src/policy/fees.cpp:
```diff
@@ -998,19 +998,34 @@ void CBlockPolicyEstimator::FlushUnconfirmed() {
+                                   double fee_filter_spacing)
 {
-    CAmount minFeeLimit = std::max(CAmount(1), minIncrementalFee.GetFeePerK() / 2);
+    std::set<double> feeset;
```
nit, the second commit would have been easier to review without the localvar renamings, but since you rename them, may as well also update `feeset` to `fee_set` so that they are all updated.

vasild replied:
Done, renames in a separate commit.

jonatack reviewed (commented):
ACK ecb5e10247a80356e1d22c2970e8c9b8fabff486

The first commit message should be fixed from "Co-authoread-by:" to "Co-authored-by:"

jonatack commented on src/policy/fees.cpp:
```diff
@@ -998,19 +998,34 @@ void CBlockPolicyEstimator::FlushUnconfirmed() {
-    CAmount minFeeLimit = std::max(CAmount(1), minIncrementalFee.GetFeePerK() / 2);
+    std::set<double> feeset;
+
+    CAmount min_fee_limit = std::max(CAmount(1), min_incremental_fee.GetFeePerK() / 2);
```
ecb5e10247a80356e1d22c2970e8c9b8fabff486
```suggestion
    const CAmount min_fee_limit{std::max(CAmount(1), min_incremental_fee.GetFeePerK() / 2)};
```

vasild committed:
style: rename variables to match coding style

Rename the variables that were touched by the previous commit (split
logical from style changes).

minIncrementalFee -> min_incremental_fee
minFeeLimit -> min_fee_limit
bucketBoundary -> bucket_boundary
feeset -> fee_set
FeeFilterRounder::feeset -> FeeFilterRounder::m_fee_set

vasild commented:
`ecb5e10247...b2311fe154`: address suggestions

jonatack commented:
ACK b2311fe154dbe3cdc5fac2cf65205ae25f20c8df

promag commented on src/policy/fees.cpp:
```diff
@@ -1010,7 +1010,8 @@ FeeFilterRounder::FeeFilterRounder(const CFeeRate& minIncrementalFee)
 {
     std::set<double>::iterator it = feeset.lower_bound(currentMinFee);
-    if ((it != feeset.begin() && insecure_rand.rand32() % 3 != 0) || it == feeset.end()) {
+    const auto rand = WITH_LOCK(m_insecure_rand_mutex, return insecure_rand.rand32());
```
adfd07564d5544f327d7432faddcf9612a6a22b3

Just noting that now now `rand32()` is always called.

ajtowns replied:
Could have 

```c++
    bool PreferLower() { LOCK(m_insecure_rand_mutex); return insecure_rand.rand32() % 3 == 0; }

    if (it != m_fee_set.begin() && PreferLower() || it == m_fee_set.end()) --it;
```

vasild replied:
Done, thanks!

promag reviewed (approved):
Code review ACK b2311fe154dbe3cdc5fac2cf65205ae25f20c8df

vasild commented:
`b2311fe154...f69670bfc2`: call `rand32()` only if necessary

Invalidates ACKs from @jonatack, @promag

jonatack commented on src/policy/fees.cpp:
```diff
@@ -998,20 +998,38 @@ void CBlockPolicyEstimator::FlushUnconfirmed() {
+        LOCK(m_insecure_rand_mutex);
+        return insecure_rand.rand32() % 3 != 0;
+    };
+    if (it == m_fee_set.end() || (it != m_fee_set.begin() && PreferLower())) {
```
(This alternative is more straightforward to my eye. Feel free to ignore.)
```diff
-    const auto PreferLower = [this]() -> bool {
-        LOCK(m_insecure_rand_mutex);
-        return insecure_rand.rand32() % 3 != 0;
-    };
-    if (it == m_fee_set.end() || (it != m_fee_set.begin() && PreferLower())) {
+    if (it == m_fee_set.end() || (it != m_fee_set.begin() && WITH_LOCK(m_insecure_rand_mutex, return insecure_rand.rand32()) % 3) {
```

Ahmadchurra681 replied:
Legally help for the safety of worlds

ajtowns replied:
You already have a class just for this logic; seems weird to use a lambda when you could just have a dedicated method? (Having a separate function makes the `if` line easier to read IMO -- the github comment already requires scrolling) *shrug*

vasild replied:
Alright, dropped the function and used `WITH_LOCK()`. I wrapped the line coz long lines arehardtoreadandrequirescrolling in someenvironments (e.g.github) and also when a single piece of it is changed, then the whole line appears as changed in diffs - see the diff at the bottom of https://github.com/bitcoin/bitcoin/pull/15936#discussion_r850190114 for example.

In other words, I agree it is better to be purple, but I like the blueish tint.

jonatack replied:
Making review more difficult on the platform of a trusted third party could be considered a feature :smiling_imp:

jonatack reviewed (commented):
ACK  f69670bfc2b62c5e59b559726bf16c842e9f0f2e

vasild commented:
`f69670bfc2...8173f160e0`: rebase and address [suggestion](https://github.com/bitcoin/bitcoin/pull/24407#discussion_r851239444).

jonatack commented:
re-ACK 8173f160e085186c9bcc7f3506205c309ee66af6

promag reviewed (approved):
Code review ACK 8173f160e085186c9bcc7f3506205c309ee66af6

laanwj commented:
Code review ACK 8173f160e085186c9bcc7f3506205c309ee66af6

mzumsande commented:
CI currently fails on master, probably in connection with this:
```
policy/fees.cpp:1042:10: error: calling function 'MaybeCheckNotHeld' requires negative capability '!m_insecure_rand_mutex' [-Werror,-Wthread-safety-analysis]
         WITH_LOCK(m_insecure_rand_mutex, return insecure_rand.rand32()) % 3 != 0)) {
```
(https://github.com/bitcoin/bitcoin/runs/8876014446)
